File: 0229-majority-element-ii/0229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> result;
        int n=nums.size();
        int cnt1=0,cnt2=0,el1=INT_MIN,el2=INT_MIN;
        for(int i=0;i<n;i++)
        {
            if(cnt1==0 && el2!=nums[i]) {cnt1=1;el1=nums[i];}
            else if(cnt2==0 && el1!=nums[i]) {cnt2=1;el2=nums[i];}
            else if(nums[i]==el1) cnt1++;
            else if(nums[i]==el2) cnt2++;
            else {cnt1--;cnt2--;};
        }
        
        int el1C=0,el2C=0;
        for(int i=0;i<n;i++)
        {
            if(el1==nums[i]) el1C++;
            if(el2==nums[i]) el2C++;
        }
        if(el1C>n/3) result.push_back(el1);
        if(el2C>n/3) result.push_back(el2);
        return result;
     
    }
};

File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) 
    {
        int ct1=0,ct2=0,el1=0,el2=0;
        
        for(int i=0;i<nums.length;i++)
        {
            if(ct1 == 0 && nums[i]!=el2) {el1=nums[i];ct1++;}
            
            else if(ct2==0 && nums[i]!=el1) {el2=nums[i];ct2++;}
            
            else if(el1==nums[i]) ct1++;
            
            else if (el2 == nums[i] ) ct2++;
            
            else {ct1--;ct2--;}
        }
        
        List<Integer> result = new ArrayList<>();
        
        ct1=0;
        ct2=0;
        
        for(int i=0;i<nums.length;i++)
        {
            if(el1==nums[i]) ct1++;
            if(el2==nums[i]) ct2++;
        }
        
        if(ct1>(nums.length/3)) result.add(el1);
        if(ct2>(nums.length/3) && el1!=el2) result.add(el2);
        
        return result;
    }
}